frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/NodeVersionDetector.java

```diff
@@ -0,0 +1,160 @@
+package com.github.eirslett.maven.plugins.frontend.lib;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Objects.isNull;
+import static java.util.Optional.empty;
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class NodeVersionDetector {
+
+    private static final String TOOL_VERSIONS_FILENAME = ".tool-versions";
+
+    public static String getNodeVersion(File workingDir, String providedNodeVersion, String genericNodeVersionFile) throws Exception {
+        Logger logger = getLogger(NodeVersionDetector.class);
+
+        if (!isNull(providedNodeVersion) && !providedNodeVersion.trim().isEmpty()) {
+            logger.debug("Looks like a node version was set so using that: " + providedNodeVersion);
```

Comment:
Could/should we issue a warning if the version set in pom.xml and .npmrc etc are mismatched?

Reply:
Great idea, I thought about this too. It creates a new set of problems though.

1. They may intentionally mismatch. With aliases like 'latest-<codename>' one might be supported but not the other. Similarly, this could happen with something like just the major like '20' in the version manager file. Then there needs to be a way to disable the warning.
2. What happens when there's multiple version manager files?
3. It creates more complicated code to read, finding the right factoring for this would make more sense after we've added the rest of the node version manager support.

All things that can be solved, but left out of scope because of the amount of work it entails. I don't think the value of it is very high. Particularly because this PR alone theoretically puts everyone in DC on a path to have a single file defining it and not having anything in the POM with a long-term self-correction incentive for engineers (Renovate).